### PR TITLE
Cs FX enhancement - Polling commands

### PR DESCRIPTION
**Pull request by ChanochShayner**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/45506

## Description
Added the *polling* argument to the following commands:
- ***cs-fx-upload-file*** 
- ***cs-fx-submit-uploaded-file*** 
- ***cs-fx-submit-url***
When *polling* is set to true, the command will try to return the results.
The *polling* argument enables users to search CrowdStrike using a single command, that doesn't require **GenericPolling Playbook**. 
- Added *extended_data* argument to the ***cs-fx-get-full-report*** command, when set to true the report will return with extended data which includes mitre attacks and signature information.

## Does it break backward compatibility?
- [x] No

## Must have
- [x] Tests

### CODE REVIEW

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 1]
```diff
@@ -835,8 +898,161 @@ def find_submission_id_command(
...
+        is_polling = True
+    elif command == 'cs-fx-submit-uploaded-file' and args.get('polling'):
+        is_polling = True
+    elif command == 'cs-fx-submit-url' and args.get('polling'):
+        is_polling = True
+    return is_polling
+
+
+def validate_command_args(command, args):
+    if 'ids' in args:
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:1022]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 2]
```diff
@@ -860,8 +1081,17 @@ def main():
             'cs-fx-find-reports': find_sandbox_reports_command,
             'cs-fx-find-submission-id': find_submission_id_command
         }
-        if command in commands:
-            return_outputs(*commands[command](client, **demisto.args()))  # type: ignore[operator]
+        if command in polling_commands:
+            if should_run_command_as_polling(command, args):
+                validate_command_args(command, args)
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:1086]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 3]
```diff
@@ -835,8 +898,161 @@ def find_submission_id_command(
...
+    return run_polling_command(client, args, 'cs-fx-submit-uploaded-file',
+                               send_uploaded_file_to_sandbox_analysis_command, get_full_report_command, 'FILE')
+
+
+def submit_uploaded_url_polling_command(client, args):
+    return run_polling_command(client, args, 'cs-fx-submit-url', send_url_to_sandbox_analysis_command,
+                               get_full_report_command, 'URL')
+
+
+def should_run_command_as_polling(command, args):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:1010]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 4]
```diff
@@ -835,8 +898,161 @@ def find_submission_id_command(
...
+        args.pop('submit_file')
+    if 'enable_tor' in args:
+        args.pop('enable_tor')
+    if 'interval_in_seconds' in args:
+        args.pop('interval_in_seconds')
+    if 'polling' in args:
+        args.pop('polling')
+
+
+def run_polling_command(client, args: dict, cmd: str, upload_function: Callable, results_function: Callable, item_type):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:933]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 5]
```diff
@@ -835,8 +898,161 @@ def find_submission_id_command(
     return tableToMarkdown("CrowdStrike Falcon X response:", filtered_outputs), entry_context, [response]
 
 
+def get_results_function_args(outputs, extended_data, item_type):
+    if isinstance(outputs, list):
+        outputs = outputs[0]
+
+    results_function_args: dict = {}
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:905]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 6]
```diff
@@ -494,10 +496,18 @@ def parse_outputs(
             if api_res_resources and sandbox:
                 api_res_sandbox = sandbox[0]
                 sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, sandbox_fields)
+                extra_sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, extra_sandbox_fields)
+
+                if extra_sandbox_group_outputs.get('processes'):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:501]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 7]
```diff
@@ -50,10 +56,37 @@ script:
...
+      description: If set to true, the report will return extended data which includes mitre attacks and signature information.
+      isArray: false
+      name: extended_data
+      predefined:
+        - 'true'
+        - 'false'
+      required: false
+      secret: false
+    - default: false
+      defaultValue: '60'
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:85]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 8]
```diff
@@ -203,7 +803,7 @@ script:
     - contextPath: csfalconx.resource.environment_id
       description: Environment ID.
       type: String
-    - contextPath: csfalconx.resource.snadbox.environment_description
+    - contextPath: csfalconx.resource.sandbox.environment_description
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:806]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml [hunk 9]
```diff
@@ -23,15 +23,21 @@ configuration:
...
   - arguments:
     - default: false
       description: Name of the file to upload for sandbox analysis.
       isArray: false
       name: file_name
-      required: true
+      required: false
+      secret: false
+    - default: false
+      description: Used just for the inner polling flow.
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.yml:37]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 10]
```diff
@@ -835,8 +898,149 @@ def find_submission_id_command(
...
+    if 'enable_tor' in args:
+        args.pop('enable_tor')
+    if 'interval_in_seconds' in args:
+        args.pop('interval_in_seconds')
+    if 'polling' in args:
+        args.pop('polling')
+
+
+def should_run_upload_function(args):
+    return not args.get('ids')
```
Comment (guykeller) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:929]: Let's change the name + add a docstring explaining why this means what it means

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 11]
```diff
@@ -835,8 +898,149 @@ def find_submission_id_command(
...
+        results_function_args = get_results_function_args(outputs, extended_data, item_type, interval_in_secs)
+        # schedule next poll
+        scheduled_command = ScheduledCommand(
+            command=cmd,
+            next_run_in_seconds=interval_in_secs,
+            args=results_function_args,
+            timeout_in_seconds=6000)
+        command_results.scheduled_command = scheduled_command
+        return command_results
+    # not a new search, get search status
```
Comment (guykeller) [inline Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:974]:
```suggestion

    # not a new search, get search status
```

**Review by guykeller — APPROVED** [review]

**Comment by guykeller** [comment]
@David-BMS Can you please test this and make sure it fits your needs?
Thanks!

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py [hunk 12]
```diff
@@ -494,10 +496,18 @@ def parse_outputs(
             if api_res_resources and sandbox:
                 api_res_sandbox = sandbox[0]
                 sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, sandbox_fields)
+                extra_sandbox_group_outputs = add_outputs_from_dict(api_res_sandbox, extra_sandbox_fields)
+
+                if extra_sandbox_group_outputs.get('processes'):
+                    for pro in extra_sandbox_group_outputs.get('processes'):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py]: already fixed

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py (hunk 10 above)
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:929]: done

File: Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py (hunk 11 above)
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalconX/Integrations/CrowdStrikeFalconX/CrowdStrikeFalconX.py:974]: done

**Review by David-BMS — COMMENTED** [review]
Notice the id is
  id: CrowdStrike Falcon X_copy

I dont see the implementation of  hash argument for the find reports command

**Review by David-BMS — APPROVED** [review]

File: Packs/CrowdStrikeFalconX/ReleaseNotes/1_1_7.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+
+#### Integrations
+##### CrowdStrike Falcon X
+Added the *polling* argument to the following commands:
+- ***cs-fx-upload-file*** 
+- ***cs-fx-submit-uploaded-file*** 
+- ***cs-fx-submit-url***
+When *polling* is set to true, the command will try to return the results.
+The *polling* argument enables users to search CrowdStrike using a single command, that doesn't require **GenericPolling Playbook**. 
+- Added *extended_data* argument to the ***cs-fx-get-full-report*** command, when set to true the report will return with extended data which includes mitre attacks and signature information.
```
Comment (guykeller) [inline Packs/CrowdStrikeFalconX/ReleaseNotes/1_1_7.md:10]:
```suggestion
- Added the *extended_data* argument to the ***cs-fx-get-full-report*** command. When set to true the report will return with extended data which includes mitre attacks and signature information.
```

**Review by guykeller — APPROVED** [review]

**Comment by ChanochShayner** [comment]
We will need a force merge due to - https://github.com/demisto/etc/issues/26209
@guykeller @yuvalbenshalom

**Pull request merged**